### PR TITLE
Stop ands reschedule commands that block checkpoints.

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
@coleaeason @righdforsa 

I think this works now. The test I added is failing in Travis but working locally. It's a sort of useless test anyway, because without tailing the logs while you run it, you can't verify that we actually hit the case where we triggered a checkpoint. We may want to remove the test. It also takes ~5 minutes to run. We could probably improve it somehow, if we want.

Issues:
$ https://github.com/Expensify/Expensify/issues/127070

### CODE REVIEW

**Comment by tylerkaraszewski** [comment]
As of the most recent commit, we have a test that runs and fails. This is fine. The test may get removed, as the only way to verify that it does something correctly is to watch the logs, and it's a very slow test. Regardless, this is the progress here:

I added all the code to interrupt commands so that checkpoints could finish. In the latest commit, these commands just fail. Before I write the code to reschedule them to run again, I wanted to make sure we're actually seeing some benefit from the checkpointing change.

First, the above code with the interruption commented out, so that the checkpoint waits for everything to complete:
```
[checkpoint] Waited 271ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 156ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 442ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 837ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 179ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 88ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 347ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 526ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 158ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 172ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 334ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 47ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 126ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 791ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 91ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 57ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 131ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 96ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 90ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 639ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 179ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 56ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 102ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 358ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 181ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 129ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 34ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 404ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 174ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 135ms for pending transactions. Starting complete checkpoint.
```

And, the code above where we interrupt commands:
```
[checkpoint] Waited 1ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 3ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 5ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 19ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 0ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 15ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 6ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 0ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 2ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 9ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 12ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 10ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 16ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 2ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 0ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 30ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 18ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 11ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 0ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 1ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 0ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 28ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 0ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 32ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 10ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 6ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 34ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 45ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 14ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 16ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 72ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 21ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 12ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 41ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 26ms for pending transactions. Starting complete checkpoint.
[checkpoint] Waited 72ms for pending transactions. Starting complete checkpoint.
```

So, this isn't the most rigorous testing ever done, but it does look like this is working, the checkpoints are quickly unblocked after the change.

I'll continue and add the command rescheduling.

File: test/clustertest/tests/CheckpointTest.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#include "../BedrockClusterTester.h"
+
+struct CheckpointTest : tpunit::TestFixture {
+    CheckpointTest()
+        : tpunit::TestFixture("Checkpoint",
+                              BEFORE_CLASS(CheckpointTest::setup),
+                              AFTER_CLASS(CheckpointTest::teardown),
```
Comment (coleaeason) [inline test/clustertest/tests/CheckpointTest.cpp]: these seem out of logical order?

File: test/clustertest/tests/CheckpointTest.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        for (int i : {0, 1, 2}) {
+            threads.emplace_back([this, i, &totalRequestFailures, &m](){
+                BedrockTester& brtester = tester->getTester(i);
+
+                // Let's make ourselves 200 commands to spam at each node.
+                vector<SData> requests;
+                int numCommands = 200;
+                for (int j = 0; j < numCommands; j++) {
+                    // Pick a command:
+                    static vector<string> commands = {"idcollision", "slowquery", "slowprocessquery"};
```
Comment (coleaeason) [inline test/clustertest/tests/CheckpointTest.cpp]: this whole block of code is hard to read, could use some whitespace maybe

File: test/clustertest/tests/CheckpointTest.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                }
+                totalRequestFailures.fetch_add(failures);
+            });
+        }
+
+        // Done.
+        for (thread& t : threads) {
+            t.join();
+        }
+        threads.clear();
```
Comment (coleaeason) [inline test/clustertest/tests/CheckpointTest.cpp]: is this necessary?

**Pull request merged**